### ISSUE TITLE
how to ensure a redraw?

### ISSUE THREAD

**Issue by md-work** [ruby-gnome/ruby-gnome#1292]
I can't find a way to ensure the ProgressBar gets drawn before the first sleep.

Sometimes the ProgressBar gets drawn before the first sleep, sometimes not.

Thanks!

```ruby
require 'gtk3'


window = Gtk::Window.new
btn_start = Gtk::Button.new(label: 'start')
btn_start.signal_connect('clicked') do
  pbar = Gtk::ProgressBar.new
  window.remove(btn_start)
  window.add(pbar)
  window.show_all
  pbar.fraction = 0

  Gtk.main_iteration while Gtk.events_pending?
  # How to ensure the progress bar gets drawn here!?
  sleep(2)
  pbar.fraction = 0.5

  Gtk.main_iteration while Gtk.events_pending?
  sleep(2)
  pbar.fraction = 1

  window.remove(pbar)
  window.add(btn_start)
end
window.add(btn_start)
window.show_all
Gtk.main
```

**Comment by kou** [comment]
You must back to the main event loop instead of using `Gtk.main_iteration while Gtk.events_pending?`:

```ruby
#!/usr/bin/env ruby

require 'gtk3'

window = Gtk::Window.new
box = Gtk::Box.new(:vertical)
box.show
btn_start = Gtk::Button.new(label: 'start')
btn_start.show
box.add(btn_start)
pbar = Gtk::ProgressBar.new
box.add(pbar)
window.add(box)
window.show

btn_start.signal_connect('clicked') do
  btn_start.hide
  pbar.show
  pbar.fraction = 0

  GLib::Timeout.add_seconds(2) do
    pbar.fraction = 0.5

    GLib::Timeout.add_seconds(2) do
      pbar.fraction = 1

      pbar.hide
      btn_start.show

      GLib::Source::REMOVE
    end

    GLib::Source::REMOVE
  end
end

Gtk.main
```

**Comment by md-work** [comment]
@kou
Sadly I can't really use your solution. I know `GLib::Timeout.add_seconds`. But the `sleep` in my example was more of a placeholder for where I've quite work intense code in the software I'm working on.

I actually tried putting a `GLib::Timeout.add_seconds` before the `sleep`. But I didn't really liked that solution. Especially because it the amount of timeout needed to redraw the gui was behaving statistically. E.g. 0.000,005 seconds sometimes did the job and sometimes not.
```ruby
# [...]
btn_start.signal_connect('clicked') do
  btn_start.hide
  pbar.show
  pbar.fraction = 0

  GLib::Timeout.add_seconds(0.000_005) do
    Gtk.main_quit
    GLib::Source::REMOVE
  end
  Gtk.main
  # a placeholder for some useful, but load intense work - like sleep ;-)
  (0..21_000_000).to_a.reverse.sort
  pbar.fraction = 0.5
  # [...]
```

**Comment by md-work** [comment]
Found two solutions which are not as ugly as a timeout.
But I'm still not completely happy.

&nbsp;

First: `wait_for_draw`
```ruby
#!/usr/bin/env ruby

require 'gtk3'


def force_redraw(widget)
  # has to be done twice, because there might be
  # a first extra redraw for resizing the window
  2.times do
    widget.queue_draw
    widget.wait_for_draw
  end
end


window = Gtk::Window.new
btn_start = Gtk::Button.new(label: 'start')
btn_start.signal_connect('clicked') do
  pbar = Gtk::ProgressBar.new
  window.remove(btn_start)
  window.add(pbar)
  window.show_all
  pbar.fraction = 0
  force_redraw(window)

  sleep(2)  # a placeholder for some meaningful, load intense work
  pbar.fraction = 0.5
  force_redraw(window)

  sleep(2)  # a placeholder for some meaningful, load intense work
  window.remove(pbar)
  window.add(btn_start)
  force_redraw(window)
end
window.add(btn_start)
window.show_all
Gtk.main
```

&nbsp;

Second: `draw` event
```ruby
require 'gtk3'


def force_redraw
  $redraw = true
  Gtk.main
end


window = Gtk::Window.new
window.signal_connect('draw') do
  if $redraw
    $redraw = false
    Gtk.main_quit
  end
end
btn_start = Gtk::Button.new(label: 'start')
btn_start.signal_connect('clicked') do
  pbar = Gtk::ProgressBar.new
  window.remove(btn_start)
  window.add(pbar)
  window.show_all
  pbar.fraction = 0
  force_redraw

  sleep(0.5)  # a placeholder for some useful, but load intense work
  pbar.fraction = 0.5
  force_redraw

  sleep(0.5)  # a placeholder for some useful, but load intense work
  window.remove(pbar)
  window.add(btn_start)
  force_redraw
end
window.add(btn_start)
window.show_all
Gtk.main
```

**Comment by kou** [comment]
You must not run heavy task in callback.

You can use `Thread` but you must not touch any GTK objects outside of the main thread. You can run any code in the main thread by `GLib::Idle.add`:

```ruby
#!/usr/bin/env ruby

require 'gtk3'

window = Gtk::Window.new
box = Gtk::Box.new(:vertical)
box.show
btn_start = Gtk::Button.new(label: 'start')
btn_start.show
box.add(btn_start)
pbar = Gtk::ProgressBar.new
box.add(pbar)
window.add(box)
window.show

btn_start.signal_connect('clicked') do
  btn_start.hide
  pbar.show
  pbar.fraction = 0

  Thread.new do
    sleep(2)
    GLib::Idle.add do
      pbar.fraction = 0.5
      GLib::Source::REMOVE
    end
    sleep(2)
    GLib::Idle.add do
      pbar.fraction = 1
      GLib::Source::REMOVE
    end
    sleep(2)
    GLib::Idle.add do
      pbar.hide
      btn_start.show
    end
  end
end

Gtk.main
```

**Comment by md-work** [comment]
> You must not run heavy task in callback.
> 
> You can use `Thread` but you must not touch any GTK objects outside of the main thread. You can run any code in the main thread by `GLib::Idle.add`:

Actually I already had some thoughts about it and read the [thread tips](https://ruby-gnome2.osdn.jp/hiki.cgi?tips_threads).

Currently I'm trying to avoid using Thread, because of the architecture overhead*.
And for the moment I'm happy with what I can do without Thread - except the ensure redraw stuff. But I may be fine with the `signal_connect('draw')` solution.

\* I know it's manageable. But currently less is more for me.

&nbsp;
---
&nbsp;

I guess even when running multi threaded, sometimes you might want to force the Gtk thread to redraw as fast as possible or at least being notified after redraw.

So maybe someone knows how to improve the solutions I found (or maybe knows another better, but lightweight solution).

&nbsp;
The open questions I have are:
* How can I check if a redraw is needed currently?
  * In other words: Is the screen content consistent with the configured widgets?
* How can I ensure, that I called `wait_for_draw` as many times as needed to fully redraw.
  * Looks like `wait_for_draw` needs to be called at least twice if the window is being resized and redrawn, because the contents needs more space then before.

**Comment by kou** [comment]
As I said that your approach (heavy task and wait for redraw in callback) isn't the GTK way (normally other GUI toolkits don't like the approach).

If you want to get more feedback, https://discourse.gnome.org/ will be a better place.

**Comment by kojix2** [comment]
I'm not a professional programmer and may not fully understand what you say, but if you're creating a GUI application using _only_ Ruby, I think `druby` can be useful. druby is a bit old tool, but it still works. You can easily send and receive Ruby objects between multiple ruby processes via druby server.
* https://ruby-doc.org/stdlib-2.6.4/libdoc/drb/rdoc/DRb.html
* http://www.druby.org/sidruby/the-druby-book.html

This will make it a little easier to avoid using Ruby threads and perform asynchronous processing.
